vpr/src/base/netlist_walker.h

```diff
@@ -25,16 +25,24 @@ class NetlistVisitor {
         virtual ~NetlistVisitor() = default;
         void start() { start_impl(); }
         void visit_top(const char* top_level_name) { visit_top_impl(top_level_name); }
-        void visit_clb(const t_pb* clb) { visit_clb_impl(clb); }
+        void visit_clb(ClusterBlockId blk_id, const t_pb* clb) { visit_clb_impl(blk_id, clb); }
         void visit_atom(const t_pb* atom) { visit_atom_impl(atom); }
+        void visit_open(const t_pb* atom) { visit_open_impl(atom); }
+        void visit_all(const t_pb_routes &top_pb_route, const t_pb* pb, const t_pb_graph_node* pb_graph_node) {
+                VTR_ASSERT(pb == nullptr || pb_graph_node == pb->pb_graph_node);
+                VTR_ASSERT(pb_graph_node != nullptr);
+                visit_all_impl(top_pb_route, pb, pb_graph_node);
+        }
         void finish() { finish_impl(); }
 
     protected:
         //All implementation methods are no-ops in this base class
         virtual void start_impl();
         virtual void visit_top_impl(const char* top_level_name);
-        virtual void visit_clb_impl(const t_pb* clb);
+        virtual void visit_clb_impl(ClusterBlockId blk_id, const t_pb* clb);
         virtual void visit_atom_impl(const t_pb* atom);
+        virtual void visit_open_impl(const t_pb* atom);
+        virtual void visit_all_impl(const t_pb_routes &top_pb_route, const t_pb* pb, const t_pb_graph_node* pb_graph_node);
```

Comment:
Probably worth commenting briefly about the relation ship between these arguments, so someone implementing this function doesn't have to go figure out the invariants on these data structures themselves.

I think pb/pb_graph_node correspond to the current hierarchy instance being traversed, but top_pb_route corresponds to the top-level block's entire routing.

Reply:
Sure, done!